src/hash/blake3_u4_compact.rs

```diff
@@ -0,0 +1,479 @@
+                }
+            ),
+            1,
+            false,
+            0,
+            &format!("unpack msg{}p0", i),
+        );
+
+        // handle padding if it is the last block
+        if i == (num_blocks - 1) && msg_len != 64 {
+            // due to LE representation, msg portion can be on top of padding.
+            let j = msg_len % 4;
+            let pad_bytes = 64 + j - msg_len - 4;
+
+            stack.custom(
+                script!(
+                    //Drop whatever padding has been added for packing to limbs and pad with zeros
+                    for _ in 0..(pad_bytes*2){
+                        OP_DROP
+                    }
+
+                    for _ in 0..(j*2){
+                        OP_TOALTSTACK
+                    }
+
+                    for _ in 0..(4-j) * 2{
+                        OP_DROP
+                    }
+
+                    for _ in 0..(4-j) * 2{
+                        OP_0
+                    }
+
+                    for _ in 0..(j*2){
+                        OP_FROMALTSTACK
+                    }
+
+                    for _ in 0..(pad_bytes*2){
+                        OP_0
+                    }
+                ),
+                0,
+                false,
+                0,
+                "padding",
+            );
+        }
+
+        //make a hashmap of msgs
+        let mut original_message = Vec::new();
+        for i in 0..16 {
+            let m = stack.define(8, &format!("msg_{}", i));
+            original_message.push(m);
+        }
+
+        // create the current block message map
+        let mut message = HashMap::new();
+        for m in 0..16 {
+            message.insert(m as u8, original_message[m as usize]);
+        }
+
+        compress(
+            stack,
+            i != 0,
+            0,
+            msg_len.min(64),
+            get_flags_for_block(i, num_blocks),
+            message,
+            &tables,
+            8,
+            i == num_blocks - 1,
+        );
+
+        //delete the intermediate states
+        for _ in 0..8 {
+            stack.drop(stack.get_var_from_stack(0));
+        }
+
+        if msg_len > 64 {
+            msg_len -= 64;
+        }
+    }
+    // drop tables
+    tables.drop(stack);
+
+    // get the result hash
+    stack.from_altstack_joined(8 as u32 * 8, "blake3-hash");
+}
+
+#[cfg(test)]
+mod tests {
+
+    pub use bitcoin_script::script;
+    use bitcoin_script_stack::{debugger::debug_script, optimizer::optimize, stack::StackTracker};
+    use rand::Rng;
+
+    use super::*;
+    use crate::{execute_script, u4::u4_std::u4_hex_to_nibbles};
+
+    fn add_padding(input: String) -> String {
+        let len_bytes = input.len() / 2;
+        let mut res = String::from(input);
+
+        if len_bytes % 64 != 0 {
+            res.push_str(&"0".repeat((64 - (len_bytes % 64)) * 2)); //zero should be added as padding but this is done intentionally to test if blake3_u4_compact can handle wrong padding
```

Comment:
Isn't this doing correct padding, not incorrect padding as suggested by the comment?